graph/path/yen_ksp.go

```diff
@@ -0,0 +1,156 @@
+// Copyright ©2018 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package path
+
+import (
+	"sort"
+
+	"gonum.org/v1/gonum/graph"
+	"gonum.org/v1/gonum/graph/internal/set"
+)
+
+func YenKShortestPath(g graph.Graph, k int, s, t graph.Node) [][]graph.Node {
+	yk := yenKSPAdjuster{
+		g:       g,
+		visited: make(map[int64]set.Int64s),
+	}
+
+	var weight Weighting
+	if wg, ok := g.(Weighted); ok {
+		weight = wg.Weight
+	} else {
+		weight = UniformCost(g)
+	}
+
+	yk.weight = weight
+
+	var paths [][]graph.Node
+
+	shortest, _ := DijkstraFrom(s, yk).To(t.ID())
+	if len(shortest) != 0 {
+		paths = append(paths, shortest)
+	} else {
+		return paths
+	}
+
+	var pot []yenShortest
+
+	for i := int64(1); i < int64(k); i++ {
+		for n := 0; n < (len(paths[i-1]) - 1); n++ {
+			spur := paths[i-1][n]
+			root := make([]graph.Node, len(paths[i-1][:n+1]))
+			copy(root, paths[i-1][:n+1])
+
+			var rootWeight float64
+			for x := 1; x < len(root); x++ {
+				w, _ := weight(root[x-1].ID(), root[x].ID())
+				rootWeight += w
+			}
+
+			for _, path := range paths {
+				if len(path) > n {
+					ok := true
+					for x := 0; x < len(root); x++ {
+						if path[x].ID() != root[x].ID() {
+							ok = false
+							break
+						}
+					}
+					if ok {
+						yk.addVisited(path[n].ID(), path[n+1].ID())
+					}
+				}
+			}
+
+			spath, weight := DijkstraFrom(spur, yk).To(t.ID())
+			size := len(root) - 1
```

Comment:
`size` is only used in the first condition branch below (and only once), so just write `nroot := root[:len(root)-1]`. In fact, `nroot` and `root` are the same and `root` is discarded at the end of the iteration (more later), so the condition below becomes:

```
if len(root) > 1 {
    root = append(root[:len(root)-1], spath...)
    pot = append(pot, yenShortest{root, weight + rootWeight})
} else {
    pot = append(pot, yenShortest{spath, weight})
}
```